GVRf/Framework/framework/src/main/jni/engine/picker/picker_jni.cpp

```diff
@@ -25,38 +25,77 @@
 
 namespace gvr {
 extern "C" {
-JNIEXPORT jlongArray JNICALL
-Java_org_gearvrf_NativePicker_pickScene(JNIEnv * env,
-        jobject obj, jlong jscene, jfloat ox, jfloat oy, jfloat z, jfloat dx,
-        jfloat dy, jfloat dz);
-JNIEXPORT jfloat JNICALL
-Java_org_gearvrf_NativePicker_pickSceneObject(JNIEnv * env,
-        jobject obj, jlong jscene_object, jlong jcamera_rig);
-JNIEXPORT jfloatArray JNICALL
-Java_org_gearvrf_NativePicker_pickSceneObjectAgainstBoundingBox(JNIEnv * env,
-        jobject obj, jlong jscene_object, jfloat ox, jfloat oy, jfloat oz, jfloat dx,
-        jfloat dy, jfloat dz);
+    JNIEXPORT jlongArray JNICALL
+    Java_org_gearvrf_NativePicker_pickScene(JNIEnv * env,
+            jobject obj, jlong jscene, jfloat ox, jfloat oy, jfloat z, jfloat dx,
+            jfloat dy, jfloat dz);
+    JNIEXPORT jobjectArray JNICALL
+    Java_org_gearvrf_NativePicker_pickObjects(JNIEnv * env,
+            jobject obj, jlong jscene, jfloat ox, jfloat oy, jfloat oz, jfloat dx,
+            jfloat dy, jfloat dz);
+    JNIEXPORT jfloat JNICALL
+    Java_org_gearvrf_NativePicker_pickSceneObject(JNIEnv * env,
+            jobject obj, jlong jscene_object, jlong jcamera_rig);
+    JNIEXPORT jfloatArray JNICALL
+    Java_org_gearvrf_NativePicker_pickSceneObjectAgainstBoundingBox(JNIEnv * env,
+            jobject obj, jlong jscene_object, jfloat ox, jfloat oy, jfloat oz, jfloat dx,
+            jfloat dy, jfloat dz);
 }
 
 JNIEXPORT jlongArray JNICALL
 Java_org_gearvrf_NativePicker_pickScene(JNIEnv * env,
         jobject obj, jlong jscene, jfloat ox, jfloat oy, jfloat oz, jfloat dx,
         jfloat dy, jfloat dz) {
     Scene* scene = reinterpret_cast<Scene*>(jscene);
-    std::vector<EyePointeeHolder*> eye_pointee_holders =
-            Picker::pickScene(scene, ox, oy, oz, dx, dy, dz);
-    std::vector<jlong> long_eye_pointee_holders;
-    for (auto it = eye_pointee_holders.begin(); it != eye_pointee_holders.end();
-            ++it) {
-        jlong eye_pointee_holder = reinterpret_cast<jlong>(*it);
-        long_eye_pointee_holders.push_back(eye_pointee_holder);
+    std::vector<ColliderData> colliders;
+    Picker::pickScene(scene, colliders, ox, oy, oz, dx, dy, dz);
+    std::vector<jlong> long_colliders;
+    for (auto it = colliders.begin(); it != colliders.end(); ++it) {
+        const ColliderData& data = *it;
+        jlong collider = reinterpret_cast<jlong>(data.ColliderHit);
+        long_colliders.push_back(collider);
+    }
+    jlongArray jcolliders = env->NewLongArray(
+            long_colliders.size());
+    env->SetLongArrayRegion(jcolliders, 0,
+            long_colliders.size(),
+            reinterpret_cast<jlong*>(long_colliders.data()));
+    colliders.clear();
```

Comment:
I figured out how to get rid of long_colliders. That was a stupid intermediate array. I can directly store the pointers into the java array.

From: rahul [mailto:notifications@github.com]
Sent: Tuesday, June 21, 2016 10:58 PM
To: Samsung/GearVRf <GearVRf@noreply.github.com>
Cc: Nola Donato <nola.donato@samsung.com>; Author <author@noreply.github.com>
Subject: Re: [Samsung/GearVRf] Picking Reform (#621)


In GVRf/Framework/framework/src/main/jni/engine/picker/picker_jni.cpp<https://github.com/Samsung/GearVRf/pull/621#discussion_r67998207>:

> -        jlong eye_pointee_holder = reinterpret_cast<jlong>(*it);

> -        long_eye_pointee_holders.push_back(eye_pointee_holder);

> +    std::vector<ColliderData> colliders;

> +    Picker::pickScene(scene, colliders, ox, oy, oz, dx, dy, dz);

> +    std::vector<jlong> long_colliders;

> +    for (auto it = colliders.begin(); it != colliders.end(); ++it) {

> +        const ColliderData& data = *it;

> +        jlong collider = reinterpret_cast<jlong>(data.ColliderHit);

> +        long_colliders.push_back(collider);

> +    }

> +    jlongArray jcolliders = env->NewLongArray(

> +            long_colliders.size());

> +    env->SetLongArrayRegion(jcolliders, 0,

> +            long_colliders.size(),

> +            reinterpret_cast<jlong*>(long_colliders.data()));

> +    colliders.clear();

Small optimization - long_colliders.clear() here.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Samsung/GearVRf/pull/621/files/1ab966aa2aedda0d8087c5a70f72fcbe50109b1c#r67998207>, or mute the thread<https://github.com/notifications/unsubscribe/AQw1S94ub3RO90k7HjdPe-z7kgO3xIymks5qOM77gaJpZM4I6MsB>.